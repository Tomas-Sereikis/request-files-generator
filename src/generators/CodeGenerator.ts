import {
  ISwaggerScheme,
  ISwaggerSchemeArray,
  ISwaggerSchemeBoolean,
  ISwaggerSchemeNumber,
  ISwaggerSchemeObject,
  ISwaggerSchemeString,
} from '../swagger'

export interface CodeGenerator {
  of(scheme: ISwaggerScheme): string
  ofArray(scheme: ISwaggerSchemeArray): string
  ofBoolean(scheme: ISwaggerSchemeBoolean): string
  ofNumber(scheme: ISwaggerSchemeNumber): string
  ofObject(scheme: ISwaggerSchemeObject): string
  ofString(scheme: ISwaggerSchemeString): string
}
