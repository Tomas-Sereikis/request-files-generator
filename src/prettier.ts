import * as prettier from 'prettier'
import { trim } from 'lodash'

const prettierrc = require('../.prettierrc.json')

export function formatCode(code: string): string {
  return trim(
    prettier.format(code, {
      parser: 'typescript',
      ...prettierrc,
    }),
    '\n',
  )
}
