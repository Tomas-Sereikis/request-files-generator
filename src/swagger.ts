import * as swaggerParser from 'swagger-parser'
import * as swaggerScheme from 'swagger-schema-official'
import { assign, flatten } from 'lodash'

export async function validate(docs: swaggerScheme.Spec): Promise<ISwaggerDocs> {
  return await swaggerParser.validate(docs)
}

export function swaggerSchemeObjectPropertiesMerge(object: ISwaggerSchemeObject): ISwaggerSchemeObjectProperties {
  const type = SwaggerSchemeType.object
  if (object.allOf) {
    const required = flatten(object.allOf.map(curr => curr.required))
    const properties = assign({} as ISwaggerSchemeObjectProperty, ...object.allOf.map(curr => curr.properties))
    return { properties, required, type }
  } else if (object.properties && object.required) {
    return { properties: object.properties, required: object.required, type }
  } else {
    return { properties: {}, required: [], type }
  }
}

export enum SwaggerSchemeType {
  object = 'object',
  boolean = 'boolean',
  array = 'array',
  string = 'string',
  number = 'number',
  integer = 'integer',
  interface = 'interface',
}

export enum SwaggerParameterIn {
  query = 'query',
  path = 'path',
  body = 'body',
}

export interface ISwaggerDocs {
  info: ISwaggerDocsInfo
  paths: ISwaggerPaths
}

export interface ISwaggerDocsInfo {
  title: string
  description?: string
}

export interface ISwaggerPaths {
  [url: string]: ISwaggerPath
}

export interface ISwaggerPath {
  [method: string]: ISwaggerPathInfo
}

export interface ISwaggerPathInfo {
  description?: string
  summary: string
  operationId: string
  parameters?: Array<ISwaggerParameterDefaults & ISwaggerSchemeObject>
  tags: string[]
  responses: ISwaggerPathResponses
}

export interface ISwaggerParameterDefaults {
  name: string
  in: SwaggerParameterIn
  schema?: ISwaggerScheme
}

export interface ISwaggerPathResponses {
  [statusCode: string]: ISwaggerPathResponse
}

export interface ISwaggerPathResponse {
  schema: ISwaggerScheme
  description: string
}

export interface ISwaggerSchemeObjectProperty {
  [key: string]: ISwaggerScheme
}

export interface ISwaggerSchemeObjectProperties {
  properties: ISwaggerSchemeObjectProperty
  required: string[]
  type: SwaggerSchemeType.object
}

export interface ISwaggerSchemeObject extends Partial<ISwaggerSchemeObjectProperties> {
  type: SwaggerSchemeType.object
  allOf?: ISwaggerSchemeObjectProperties[]
}

export interface ISwaggerSchemeBoolean {
  type: SwaggerSchemeType.boolean
}

export interface ISwaggerSchemeArray {
  type: SwaggerSchemeType.array
  items: ISwaggerScheme
}

export interface ISwaggerSchemeString {
  type: SwaggerSchemeType.string
  enum?: string[]
}

export interface ISwaggerSchemeNumber {
  type: SwaggerSchemeType.number | SwaggerSchemeType.integer
}

export type ISwaggerScheme =
  | ISwaggerSchemeObject
  | ISwaggerSchemeBoolean
  | ISwaggerSchemeArray
  | ISwaggerSchemeString
  | ISwaggerSchemeNumber
