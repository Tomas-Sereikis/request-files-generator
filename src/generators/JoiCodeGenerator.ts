import { CodeGenerator } from './CodeGenerator'
import {
  ISwaggerScheme,
  ISwaggerSchemeArray,
  ISwaggerSchemeBoolean,
  ISwaggerSchemeNumber,
  ISwaggerSchemeObject,
  ISwaggerSchemeString,
  SwaggerSchemeType,
  joinSwaggerSchemeObjectOfAll,
} from '../swagger'
import { formatCode } from '../prettier'

export class JoiCodeGenerator implements CodeGenerator {
  of(scheme: ISwaggerScheme): string {
    const type = scheme.type
    switch (scheme.type) {
      case SwaggerSchemeType.array:
        return this.ofArray(scheme)
      case SwaggerSchemeType.boolean:
        return this.ofBoolean(scheme)
      case SwaggerSchemeType.number:
      case SwaggerSchemeType.integer:
        return this.ofNumber(scheme)
      case SwaggerSchemeType.object:
        return this.ofObject(scheme)
      case SwaggerSchemeType.string:
        return this.ofString(scheme)
      default:
        throw new Error(`Not implemented typeOf ${type}!`)
    }
  }

  ofArray(scheme: ISwaggerSchemeArray): string {
    const arrayItemsValue = this.of(scheme.items)
    const defaultValue = `joi.array().items(${arrayItemsValue})`
    return formatCode(defaultValue)
  }

  ofBoolean(scheme: ISwaggerSchemeBoolean): string {
    const defaultValue = 'joi.boolean()'
    return formatCode(defaultValue)
  }

  ofNumber(scheme: ISwaggerSchemeNumber): string {
    const defaultValue = 'joi.number()'
    return formatCode(defaultValue)
  }

  ofObject(scheme: ISwaggerSchemeObject): string {
    const info = joinSwaggerSchemeObjectOfAll(scheme)
    const keyValues = Object.keys(info.properties).reduce((previousValue, currentValue) => {
      const required = info.required.includes(currentValue) ? '.required()' : ''
      const nextScheme = this.of(info.properties[currentValue])
      return `${previousValue}'${currentValue}':${nextScheme}${required},\n`
    }, '')
    const defaultValue = `joi.object().keys({\n${keyValues}})`
    return formatCode(defaultValue)
  }

  ofString(scheme: ISwaggerSchemeString): string {
    const defaultValue =
      scheme.enum && scheme.enum.length
        ? `joi.string().valid(${JSON.stringify(scheme.enum)})`
        : `joi.string().allow('')`
    return formatCode(defaultValue)
  }
}
