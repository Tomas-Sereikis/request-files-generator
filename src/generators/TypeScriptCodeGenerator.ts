import { CodeGenerator } from './CodeGenerator'
import {
  ISwaggerScheme,
  ISwaggerSchemeArray,
  ISwaggerSchemeBoolean,
  ISwaggerSchemeNumber,
  ISwaggerSchemeObject,
  ISwaggerSchemeString,
  SwaggerSchemeType,
  joinSwaggerSchemeObjectOfAll,
} from '../swagger'
import { formatCode } from '../prettier'

interface IOfInterfaceObject {
  [key: string]: ISwaggerScheme
}

export class TypeScriptCodeGenerator implements CodeGenerator {
  private static formatCodeLocal(code: string): string {
    return formatCode(`type A = ${code}`).substr(9)
  }

  of(scheme: ISwaggerScheme): string {
    const type = scheme.type
    switch (scheme.type) {
      case SwaggerSchemeType.array:
        return this.ofArray(scheme)
      case SwaggerSchemeType.boolean:
        return this.ofBoolean(scheme)
      case SwaggerSchemeType.number:
      case SwaggerSchemeType.integer:
        return this.ofNumber(scheme)
      case SwaggerSchemeType.object:
        return this.ofObject(scheme)
      case SwaggerSchemeType.string:
        return this.ofString(scheme)
      default:
        throw new Error(`Not implemented typeOf ${type}!`)
    }
  }

  ofArray(scheme: ISwaggerSchemeArray): string {
    const arrayItemsValue = this.of(scheme.items)
    const defaultValue = `Array<${arrayItemsValue}>`
    return TypeScriptCodeGenerator.formatCodeLocal(defaultValue)
  }

  ofBoolean(scheme: ISwaggerSchemeBoolean): string {
    const defaultValue = 'boolean'
    return TypeScriptCodeGenerator.formatCodeLocal(defaultValue)
  }

  ofNumber(scheme: ISwaggerSchemeNumber): string {
    const defaultValue = 'number'
    return TypeScriptCodeGenerator.formatCodeLocal(defaultValue)
  }

  ofObject(scheme: ISwaggerSchemeObject): string {
    const info = joinSwaggerSchemeObjectOfAll(scheme)
    const keyValues = Object.keys(info.properties).reduce((previousValue, currentValue) => {
      const suffix = info.required.includes(currentValue) ? '' : '?'
      const scheme = this.of(info.properties[currentValue])
      return `${previousValue}'${currentValue}'${suffix}:${scheme},\n`
    }, '')
    const defaultValue = `{\n${keyValues}}`
    return TypeScriptCodeGenerator.formatCodeLocal(defaultValue)
  }

  ofString(scheme: ISwaggerSchemeString): string {
    const defaultValue = scheme.enum ? scheme.enum.map(val => `'${val}'`).join(' | ') : 'string'
    return TypeScriptCodeGenerator.formatCodeLocal(defaultValue)
  }

  onInterface(object: IOfInterfaceObject, required: string[]): string {
    const keyValues = Object.keys(object).reduce((previousValue, currentValue) => {
      const suffix = required.includes(currentValue) ? '' : '?'
      const scheme = this.of(object[currentValue])
      return `${previousValue}'${currentValue}'${suffix}:${scheme},\n`
    }, '')
    const defaultValue = `{\n${keyValues}}`
    return TypeScriptCodeGenerator.formatCodeLocal(defaultValue)
  }
}
