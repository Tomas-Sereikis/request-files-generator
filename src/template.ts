import * as assert from 'assert'
import { JoiCodeGenerator } from './generators/JoiCodeGenerator'
import {
  ISwaggerPathInfo,
  ISwaggerPaths,
  ISwaggerSchemeObject,
  ISwaggerSchemeObjectProperty,
  SwaggerParameterIn,
  SwaggerSchemeType,
} from './swagger'
import { TypeScriptCodeGenerator } from './generators/TypeScriptCodeGenerator'
import { camelCase, upperFirst, template } from 'lodash'
import { formatCode } from './prettier'

export interface ICreateTemplateParams {
  paths: ISwaggerPaths
  url: string
  method: string
  statusCode: string
  template: string
}

const removeNewLines = (str: string) => str.replace(/\r?\n|\r/g, ' ')

function parametersInToObjectScheme(
  parameters: ISwaggerPathInfo['parameters'],
  inType: SwaggerParameterIn,
): { count: number; required: string[]; scheme: ISwaggerSchemeObject } {
  const properties = parameters
    .filter(curr => curr.in === inType)
    .reduce(
      (previousValue, currentValue) => {
        if (
          currentValue.in === SwaggerParameterIn.body && 
          currentValue.name === 'body' && 
          currentValue.schema && 
          currentValue.schema.type === SwaggerSchemeType.object
        ) {
          return {
            ...previousValue,
            ...currentValue.schema.properties,
          }
        } else {
          return {
            ...previousValue,
            [currentValue.name]: currentValue.schema || currentValue,
          }
        }
      },
      {} as ISwaggerSchemeObjectProperty,
    )
  const required = parameters.filter(curr => curr.in === inType && curr.required).map(curr => curr.name)
  return {
    count: Object.keys(properties).length,
    required,
    scheme: {
      type: SwaggerSchemeType.object,
      properties,
      required,
    },
  }
}

function applyUrlParamsBuilder(url: string) {
  return (variable: string) => {
    const replaced = url.replace(/{(.*?)}/g, (match, value) => `\${${variable}.path.${value}}`)
    return `\`${replaced}\``
  }
}

export function createTemplate(options: ICreateTemplateParams): string {
  const tsCodeGenerator = new TypeScriptCodeGenerator()
  const joiCodeGenerator = new JoiCodeGenerator()

  const endpoint = options.paths[options.url][options.method]
  const response = endpoint ? endpoint.responses[options.statusCode] : null

  assert(!!endpoint, 'Swagger path info is not found!')
  assert(!!response, 'Swagger response is not found!')

  const bodyObjectScheme = parametersInToObjectScheme(endpoint.parameters, SwaggerParameterIn.body)
  const queryObjectScheme = parametersInToObjectScheme(endpoint.parameters, SwaggerParameterIn.query)
  const paramsObjectScheme = parametersInToObjectScheme(endpoint.parameters, SwaggerParameterIn.path)

  let optionsScheme = {}
  let required = []
  if (bodyObjectScheme.count) {
    optionsScheme = { ...optionsScheme, body: bodyObjectScheme.scheme }
    required.push(SwaggerParameterIn.body)
  }
  if (queryObjectScheme.count) {
    optionsScheme = { ...optionsScheme, query: queryObjectScheme.scheme }
  }
  if (queryObjectScheme.count && queryObjectScheme.required.length) {
    required.push(SwaggerParameterIn.query)
  }
  if (paramsObjectScheme.count) {
    optionsScheme = { ...optionsScheme, path: paramsObjectScheme.scheme }
    required.push(SwaggerParameterIn.path)
  }

  const optionsObject = tsCodeGenerator.onInterface(optionsScheme, required)
  const templateBuilder = template(options.template)
  const operationId = camelCase(endpoint.operationId)
  const description = endpoint.description ? removeNewLines(endpoint.description) : ''

  return formatCode(
    templateBuilder({
      applyUrlParams: applyUrlParamsBuilder(options.url),
      description: description,
      containsBody: ['post', 'put', 'patch'].includes(options.method),
      containsQuery: !!queryObjectScheme.count,
      responseInterface: tsCodeGenerator.of(response!.schema),
      optionsInterface: optionsObject,
      scheme: joiCodeGenerator.of(response!.schema),
      isDelete: ['delete'].includes(options.method),
      operationId,
      method: options.method,
      upperOperationId: upperFirst(operationId),
    }),
  )
}
