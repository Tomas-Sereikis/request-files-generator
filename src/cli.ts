import * as assert from 'assert'
import * as prompts from 'prompts'
import * as path from 'path'
import * as swagger from './swagger'
import { promises as fs } from 'fs'
import { createTemplate } from './template'

const MSG_OVERWRITE_FILE = 'Would you like to overwrite {fileName}'
const MSG_ENTER_TEMPLATE_PATH = 'Enter template path'
const MSG_ENTER_OUTPUT_PATH = 'Enter output path'
const MSG_ENTER_SWAGGER_JSON_URL = 'Enter swagger.json file url'
const MSG_SELECT_PATHS = 'Select paths that you would like to use'
const MSG_SELECT_OPERATIONS = 'Select operations that you would like to use'
const MSG_SELECT_STATUS_CODE = 'Select statusCode for {operationId}'

main(process.argv.slice(2)).catch(console.error)

async function main(argv: ReadonlyArray<string>) {
  const { value: templatePath } = await prompts({
    message: MSG_ENTER_TEMPLATE_PATH,
    name: 'value',
    type: 'text',
  })

  const templateStat = await fs.stat(templatePath)
  assert(templateStat.isFile(), 'Template file dose not exist!')
  const templateContent = String(await fs.readFile(templatePath))

  const { value: outputPath } = await prompts({
    message: MSG_ENTER_OUTPUT_PATH,
    name: 'value',
    type: 'text',
  })

  const outputStat = await fs.stat(outputPath)
  assert(outputStat.isDirectory(), 'Output directory dose not exist!')

  const { value: swaggerUrl } = await prompts({
    message: MSG_ENTER_SWAGGER_JSON_URL,
    name: 'value',
    type: 'text',
  })

  const docs = await swagger.validate(swaggerUrl)
  const choicesOfPaths = Object.keys(docs.paths).map(path => ({ value: path, title: path }))
  const { value: paths } = await prompts({
    message: MSG_SELECT_PATHS,
    name: 'value',
    type: 'multiselect',
    choices: choicesOfPaths,
  })

  const choicesOfOperations = Object.keys(docs.paths)
    .filter(path => paths.includes(path))
    .map(path => Object.values(docs.paths[path]))
    .reduce(
      (previousValue, currentValue) => [...previousValue, ...currentValue.map(curr => curr.operationId)],
      [] as string[],
    )
    .map(operation => ({ value: operation, title: operation }))

  const { value: operations } = await prompts({
    message: MSG_SELECT_OPERATIONS,
    name: 'value',
    type: 'multiselect',
    choices: choicesOfOperations,
  })

  for (const url of Object.keys(docs.paths)) {
    for (const method of Object.keys(docs.paths[url])) {
      const info = docs.paths[url][method]
      if (!operations.includes(info.operationId)) {
        continue
      }
      const statusCodes = Object.keys(info.responses)
      let statusCode: string | undefined
      if (statusCodes.length === 0) {
        continue
      } else if (statusCodes.length === 1) {
        statusCode = statusCodes[0]
      } else {
        const choicesOfStatusCodes = statusCodes.map(code => ({ value: code, title: code }))
        const selection = await prompts({
          message: MSG_SELECT_STATUS_CODE.replace('{operationId}', info.operationId),
          name: 'value',
          type: 'select',
          choices: choicesOfStatusCodes,
        })
        statusCode = selection.value
      }
      const template = await createTemplate({
        template: templateContent,
        paths: docs.paths,
        url,
        method,
        statusCode: statusCode!,
      })
      const filePath = path.resolve(outputPath, `${info.operationId}.ts`)
      const filePathStat = await fs.stat(filePath).catch(() => null)
      if (filePathStat && filePathStat.isFile()) {
        const { value: overwrite } = await prompts({
          message: MSG_OVERWRITE_FILE.replace('{filePath}', filePath),
          name: 'value',
          type: 'toggle',
          initial: true,
          active: 'yes',
          inactive: 'no',
        })
        if (overwrite === 'no') {
          continue
        }
      }
      await fs.writeFile(filePath, template)
    }
  }
}
