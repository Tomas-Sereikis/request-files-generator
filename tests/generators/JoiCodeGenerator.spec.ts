import * as swagger from '../../src/swagger'
import mockOfCategories from '../mocks/categories'
import mockOfPetstore from '../mocks/petstore'
import { JoiCodeGenerator } from '../../src/generators/JoiCodeGenerator'
import { SwaggerSchemeType } from '../../src/swagger'

describe('initial methods', () => {
  const joiCodeGenerator = new JoiCodeGenerator()

  describe('array builder', () => {
    test('should build array of boolean', () => {
      const value1 = joiCodeGenerator.ofArray({
        type: SwaggerSchemeType.array,
        items: { type: SwaggerSchemeType.boolean },
      })
      const value2 = joiCodeGenerator.of({
        type: SwaggerSchemeType.array,
        items: { type: SwaggerSchemeType.boolean },
      })
      expect(value1).toBe(`joi.array().items(joi.boolean())`)
      expect(value2).toBe(value2)
    })

    test('should build array of number', () => {
      const value1 = joiCodeGenerator.ofArray({
        type: SwaggerSchemeType.array,
        items: { type: SwaggerSchemeType.number },
      })
      const value2 = joiCodeGenerator.of({ type: SwaggerSchemeType.array, items: { type: SwaggerSchemeType.number } })
      expect(value1).toBe(`joi.array().items(joi.number())`)
      expect(value2).toBe(value2)
    })

    test('should build array of string', () => {
      const value1 = joiCodeGenerator.ofArray({
        type: SwaggerSchemeType.array,
        items: { type: SwaggerSchemeType.string },
      })
      const value2 = joiCodeGenerator.of({ type: SwaggerSchemeType.array, items: { type: SwaggerSchemeType.string } })
      expect(value1).toBe(`joi.array().items(joi.string().allow(''))`)
      expect(value2).toBe(value2)
    })

    test('should build array of array of boolean', () => {
      const value1 = joiCodeGenerator.ofArray({
        type: SwaggerSchemeType.array,
        items: { type: SwaggerSchemeType.array, items: { type: SwaggerSchemeType.boolean } },
      })
      const value2 = joiCodeGenerator.of({
        type: SwaggerSchemeType.array,
        items: { type: SwaggerSchemeType.array, items: { type: SwaggerSchemeType.boolean } },
      })
      expect(value1).toBe(`joi.array().items(joi.array().items(joi.boolean()))`)
      expect(value2).toBe(value2)
    })
  })

  describe('boolean builder', () => {
    test('should build boolean optional', () => {
      const value1 = joiCodeGenerator.ofBoolean({ type: SwaggerSchemeType.boolean })
      const value2 = joiCodeGenerator.of({ type: SwaggerSchemeType.boolean })
      expect(value1).toBe(`joi.boolean()`)
      expect(value1).toBe(value2)
    })
  })

  describe('number builder', () => {
    test('should build number', () => {
      const value1 = joiCodeGenerator.ofNumber({ type: SwaggerSchemeType.number })
      const value2 = joiCodeGenerator.of({ type: SwaggerSchemeType.number })
      expect(value1).toBe(`joi.number()`)
      expect(value1).toBe(value2)
    })
  })

  describe('object builder', () => {
    test('should build object with two keys number and boolean', () => {
      const value1 = joiCodeGenerator.ofObject({
        type: SwaggerSchemeType.object,
        properties: {
          a: { type: SwaggerSchemeType.number },
          b: { type: SwaggerSchemeType.boolean },
        },
        required: ['a'],
      })
      const value2 = joiCodeGenerator.of({
        type: SwaggerSchemeType.object,
        properties: {
          a: { type: SwaggerSchemeType.number },
          b: { type: SwaggerSchemeType.boolean },
        },
        required: ['a'],
      })
      expect(value1).toBe(`joi.object().keys({\n  a: joi.number().required(),\n  b: joi.boolean(),\n})`)
      expect(value1).toBe(value2)
    })

    test('should build object with two keys array and boolean', () => {
      const value1 = joiCodeGenerator.ofObject({
        type: SwaggerSchemeType.object,
        properties: {
          a: { type: SwaggerSchemeType.array, items: { type: SwaggerSchemeType.number } },
          b: { type: SwaggerSchemeType.boolean },
        },
        required: [],
      })
      const value2 = joiCodeGenerator.of({
        type: SwaggerSchemeType.object,
        properties: {
          a: { type: SwaggerSchemeType.array, items: { type: SwaggerSchemeType.number } },
          b: { type: SwaggerSchemeType.boolean },
        },
        required: [],
      })
      expect(value1).toBe(`joi.object().keys({\n  a: joi.array().items(joi.number()),\n  b: joi.boolean(),\n})`)
      expect(value1).toBe(value2)
    })
  })

  describe('string builder', () => {
    test('should build string with allow empty value', () => {
      const value1 = joiCodeGenerator.ofString({ type: SwaggerSchemeType.string })
      const value2 = joiCodeGenerator.of({ type: SwaggerSchemeType.string })
      expect(value1).toBe(`joi.string().allow('')`)
      expect(value1).toBe(value2)
    })
    test('should build string enum optional', () => {
      const value1 = joiCodeGenerator.ofString({ type: SwaggerSchemeType.string, enum: ['a', 'b'] })
      const value2 = joiCodeGenerator.of({ type: SwaggerSchemeType.string, enum: ['a', 'b'] })
      expect(value1).toBe(`joi.string().valid(['a', 'b'])`)
      expect(value1).toBe(value2)
    })
  })
})

describe('mock tests', () => {
  const joiCodeGenerator = new JoiCodeGenerator()

  test('should create categories validation', async () => {
    const docs = await swagger.validate(mockOfCategories)
    const schemeObject = docs.paths['/channel-categories'].get.responses['200'].schema
    const value = joiCodeGenerator.of(schemeObject)
    expect(value).toMatchSnapshot()
  })

  test('should create petstore pets get validation', async () => {
    const docs = await swagger.validate(mockOfPetstore)
    const schemeObject = docs.paths['/pets'].get.responses['200'].schema
    const value = joiCodeGenerator.of(schemeObject)
    expect(value).toMatchSnapshot()
  })

  test('should create petstore pet get validation', async () => {
    const docs = await swagger.validate(mockOfPetstore)
    const schemeObject = docs.paths['/pets/{id}'].get.responses['200'].schema
    const value = joiCodeGenerator.of(schemeObject)
    expect(value).toMatchSnapshot()
  })
})
