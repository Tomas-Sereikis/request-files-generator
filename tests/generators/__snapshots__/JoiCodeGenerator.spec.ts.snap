// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`mock tests should create categories validation 1`] = `
"joi.object().keys({
  data: joi
    .array()
    .items(
      joi.object().keys({
        color: joi.string().allow(''),
        id: joi
          .string()
          .allow('')
          .required(),
        region: joi
          .string()
          .allow('')
          .required(),
        sortIndex: joi.number().required(),
        svg: joi
          .string()
          .allow('')
          .required(),
        title: joi
          .string()
          .allow('')
          .required(),
      }),
    )
    .required(),
  next: joi.boolean(),
})"
`;

exports[`mock tests should create petstore pet get validation 1`] = `
"joi.object().keys({
  name: joi
    .string()
    .allow('')
    .required(),
  tag: joi.string().allow(''),
  id: joi.number().required(),
})"
`;

exports[`mock tests should create petstore pets get validation 1`] = `
"joi.array().items(
  joi.object().keys({
    name: joi
      .string()
      .allow('')
      .required(),
    tag: joi.string().allow(''),
    id: joi.number().required(),
  }),
)"
`;
